src/include/wiredtiger.in

```diff
@@ -446,6 +478,38 @@ struct __wt_cursor {
 	int __F(insert)(WT_CURSOR *cursor);
 
 	/*!
+	 * Modify an existing record.
+	 *
+	 * Both the key and value must be set and the record must already exist;
+	 * the record will be updated.
+	 *
+	 * Modification structures are applied in order, and later modifications
+	 * can update earlier modifications.
+	 *
+	 * The modify method is only supported on raw byte arrays accessed using
+	 * a WT_ITEM structure, that is, a format type of \c u.
+	 *
+	 * @snippet ex_all.c Modify an existing record
+	 *
+	 * On success, the cursor ends positioned at the modified record; to
+	 * minimize cursor resources, the WT_CURSOR::reset method should be
+	 * called as soon as the cursor no longer needs that position.
+	 *
+	 * The maximum length of a single column stored in a table is not fixed
+	 * (as it partially depends on the underlying file configuration), but
+	 * is always a small number of bytes less than 4GB.
+	 *
+	 * @param cursor the cursor handle
+	 * @param entries an array of modification data structures
+	 * @param nentries the number of modification data structures
+	 * @errors
+	 * In particular, if \c in_memory is configured for the database and
+	 * the modify requires more than the configured cache size to complete,
+	 * ::WT_CACHE_FULL is returned.
+	 */
+	int __F(modify)(WT_CURSOR *cursor, WT_MODIFY *entries, int nentries);
```

Comment:
I went looking for where we discussed this originally but couldn't find it.  That's sad because we went over this a couple of times.

The reason I don't like a new flavor of `WT_CURSOR::set_value` for this feature is that it grows the set of states that a cursor can be in (i.e., the possible combinations of cursor flags).  We'd have to think through what happens if you follow up a `set_value_partial` with something other than `WT_CURSOR::update`, such as `WT_CURSOR::insert` or `WT_CURSOR::get_value` (including for positioned and non-positioned cursors).  I'm sure we could come up with some set of rules to cover the cases, but it adds complication for no benefit that I can see.

In addition, the `WT_CURSOR::set_key` and `WT_CURSOR::set_value` methods are schema aware.  I don't believe we will ever try to make `WT_CURSOR::modify` schema aware -- it is explicitly a byte-level modification.  If we had a complex `value_format` and wanted an API to update a single field, I don't see a flavor of `WT_CURSOR::set_value` making that any easier: the way to do it would be to enable updates via a projection cursor.

> If a new value can be passed to `WT_CURSOR.modify`, why can't a new value be passed to in `WT_CURSOR.update`?

A new value can't be passed to `WT_CURSOR::modify` -- the value must exist and only a subset of bytes in the (raw) value is typically updated.  OTOH, `WT_CURSOR::update` deals in whole values, is schema aware and thus relies on `WT_CURSOR::set_value` to map from the application schema to a `WT_ITEM`.